File: 257.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> res;
        helper(res, "", root);
        return res;
    }
    
    void helper(vector<string> &res, string path, TreeNode* root){
        if(root == NULL) 
            return;
        path = path==""? to_string(root->val) : path+"->"+to_string(root->val);
        if(root->left == NULL && root->right == NULL){
            res.push_back(path);
        }
       
        helper(res, path, root->left);
        helper(res, path, root->right);
    }
};

File: golang/257/257.go
package main

import "strconv"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func dfs(root *TreeNode, str string, path *[]string) {
	if len(str) == 0 {
		str = strconv.Itoa(root.Val)
	} else {
		str = str + "->" + strconv.Itoa(root.Val)
	}
	if root.Left != nil {
		dfs(root.Left, str, path)
	}
	if root.Right != nil {
		dfs(root.Right, str, path)
	}
	if root.Left == nil && root.Right == nil {
		*path = append(*path, str)
	}
}

func binaryTreePaths(root *TreeNode) []string {
	if root == nil {
		return nil
	}
	var path []string
	dfs(root, "", &path)
	return path
}
